sympy/integrals/manualintegrate.py

```diff
@@ -64,95 +69,633 @@
 
 
-def Rule(name, props=""):
-    # GOTCHA: namedtuple class name not considered!
-    def __eq__(self, other):
-        return self.__class__ == other.__class__ and tuple.__eq__(self, other)
-    __neq__ = lambda self, other: not __eq__(self, other)
-    cls = namedtuple(name, props + " context symbol")
-    cls.__eq__ = __eq__
-    cls.__ne__ = __neq__
-    return cls
-
-ConstantRule = Rule("ConstantRule", "constant")
-ConstantTimesRule = Rule("ConstantTimesRule", "constant other substep")
-PowerRule = Rule("PowerRule", "base exp")
-NestedPowRule = Rule("NestedPowRule", "base exp")
-AddRule = Rule("AddRule", "substeps")
-URule = Rule("URule", "u_var u_func constant substep")
-PartsRule = Rule("PartsRule", "u dv v_step second_step")
-CyclicPartsRule = Rule("CyclicPartsRule", "parts_rules coefficient")
-TrigRule = Rule("TrigRule", "func arg")
-HyperbolicRule = Rule("HyperbolicRule", "func arg")
-ExpRule = Rule("ExpRule", "base exp")
-ReciprocalRule = Rule("ReciprocalRule", "func")
-ArcsinRule = Rule("ArcsinRule")
-ArcsinhRule = Rule("ArcsinhRule")
-ReciprocalSqrtQuadraticRule = Rule("ReciprocalSqrtQuadraticRule", "a b c")
-SqrtQuadraticDenomRule = Rule("SqrtQuadraticDenomRule", "a b c coeffs")
-SqrtQuadraticRule = Rule("SqrtQuadraticRule", "a b c")
-AlternativeRule = Rule("AlternativeRule", "alternatives")
-DontKnowRule = Rule("DontKnowRule")
-DerivativeRule = Rule("DerivativeRule")
-RewriteRule = Rule("RewriteRule", "rewritten substep")
-CompleteSquareRule = Rule("CompleteSquareRule", "rewritten substep")
-PiecewiseRule = Rule("PiecewiseRule", "subfunctions")
-HeavisideRule = Rule("HeavisideRule", "harg ibnd substep")
-DiracDeltaRule = Rule("DiracDeltaRule", "n a b")
-TrigSubstitutionRule = Rule("TrigSubstitutionRule",
-                            "theta func rewritten substep restriction")
-ArctanRule = Rule("ArctanRule", "a b c")
-ArctanhRule = Rule("ArctanhRule", "a b c")
-JacobiRule = Rule("JacobiRule", "n a b")
-GegenbauerRule = Rule("GegenbauerRule", "n a")
-ChebyshevTRule = Rule("ChebyshevTRule", "n")
-ChebyshevURule = Rule("ChebyshevURule", "n")
-LegendreRule = Rule("LegendreRule", "n")
-HermiteRule = Rule("HermiteRule", "n")
-LaguerreRule = Rule("LaguerreRule", "n")
-AssocLaguerreRule = Rule("AssocLaguerreRule", "n a")
-CiRule = Rule("CiRule", "a b")
-ChiRule = Rule("ChiRule", "a b")
-EiRule = Rule("EiRule", "a b")
-SiRule = Rule("SiRule", "a b")
-ShiRule = Rule("ShiRule", "a b")
-ErfRule = Rule("ErfRule", "a b c")
-FresnelCRule = Rule("FresnelCRule", "a b c")
-FresnelSRule = Rule("FresnelSRule", "a b c")
-LiRule = Rule("LiRule", "a b")
-PolylogRule = Rule("PolylogRule", "a b")
-UpperGammaRule = Rule("UpperGammaRule", "a e")
-EllipticFRule = Rule("EllipticFRule", "a d")
-EllipticERule = Rule("EllipticERule", "a d")
+@dataclass
+class Rule(ABC):
+    integrand: Expr
+    variable: Symbol
 
+    @abstractmethod
+    def eval(self) -> Expr:
+        pass
+
+    def contains_dont_know(self) -> bool:
+        return False
```

Comment:
It's usually better not to implement a method with a definite value in the superclass when we don't know what subclasses might be created in future. I would have this method raise or perhaps it should be an `abstractmethod`.

Reply:
Maybe there could be a subclass like:
```python
class SimpleRule(Rule):
    """A simple rule that does not depend on other rules"""
    def contains_dont_know(self) -> bool:
        return False
```
Then subclasses can explicitly choose whether to subclass `Rule` or `SimpleRule`.